modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java

```diff
@@ -0,0 +1,508 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+// Original code is re-licensed to Oracle by the author.
+// https://github.com/andy-goryachev/FxTextEditor/blob/master/src/goryachev/fx/FxDisconnector.java
+// Copyright © 2021-2022 Andy Goryachev <andy@goryachev.com>
+package com.sun.javafx.scene.control;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javafx.beans.InvalidationListener;
+import javafx.beans.Observable;
+import javafx.beans.value.ChangeListener;
+import javafx.beans.value.ObservableValue;
+import javafx.collections.ListChangeListener;
+import javafx.collections.MapChangeListener;
+import javafx.collections.ObservableList;
+import javafx.collections.ObservableMap;
+import javafx.collections.ObservableSet;
+import javafx.collections.SetChangeListener;
+import javafx.concurrent.Task;
+import javafx.event.Event;
+import javafx.event.EventHandler;
+import javafx.event.EventType;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.scene.control.MenuItem;
+import javafx.scene.control.SkinBase;
+import javafx.scene.control.TableColumnBase;
+import javafx.scene.control.TreeItem;
+import javafx.scene.transform.Transform;
+import javafx.stage.Window;
+
+/**
+ * This class provides convenience methods for adding various listeners, both
+ * strong and weak, as well as a single {@link #disconnect()} method to remove
+ * all listeners.
+ * <p>
+ * There are two usage patterns:
+ * <ul>
+ * <li>Client code registers a number of listeners and removes them all at once
+ * via {@link #disconnect()} call.
+ * <li>Client code registers a number of listeners and removes one via its
+ * {@link IDisconnectable} instance.
+ * </ul>
+ */
+public class ListenerHelper implements IDisconnectable {
+    private WeakReference<Object> ownerRef;
+    private final ArrayList<IDisconnectable> items = new ArrayList<>(4);
+    private static Function<SkinBase<?>,ListenerHelper> accessor;
+
+    public ListenerHelper(Object owner) {
+        ownerRef = new WeakReference<>(owner);
+    }
+
+    public ListenerHelper() {
+    }
+
+    public static void setAccessor(Function<SkinBase<?>,ListenerHelper> a) {
+        accessor = a;
+    }
+
+    public static ListenerHelper get(SkinBase<?> skin) {
+        return accessor.apply(skin);
```

Comment:
I don't think this is the case.
When this method is called, SkinBase class must be already loaded and resolved, that means it's static { } block, which sets the accessor, has been executed.  But just in case, I should move the static block in SkinBase, so we don't have any surprises.

Reply:
The `accessor` cannot be null **only if** it is used from *Skins.
In the current form, as we provide a no-argument public constructor, it is possible to create an object of `ListenerHelper` without calling setAccessor. Many unit tests added as part of this PR create such `ListenerHelper` objects.

Reply:
The `accessor` is used **only** if used from *Skins, meaning to get it from a Skin, one needs to invoke `get(SkinBase)` method, and it's magically not a null there!

It is perfectly fine to create an instance of ListenerHelper using the constructor.  The `accessor` will be null, of course, but it's ok because it will never be used until one calls `get(SkinBase)` ... and see above.

Reply:
That is my whole point. We have two paths - one for Skin classes and other for generic usage.

Approving the PR as you have added a comment regarding `ListenerHelper` class's current scope being limited to Skins. We may need to revisit the suggestion of having a derived class in future.